org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/AbstractImageBuilder.java

```diff
@@ -666,10 +756,13 @@ protected void processAnnotations(CompilationParticipantResult[] results) {
 		results[i].reset(foundAnnotations ? this.filesWithAnnotations.get(results[i].sourceFile) : null);
 	}
 
-	// even if no files have annotations, must still tell every annotation processor in case the file used to have them
-	for (CompilationParticipant participant : this.javaBuilder.participants)
-		if (participant.isAnnotationProcessor())
-			participant.processAnnotations(results);
+	boolean isEcjUsed = Compiler.class.equals(this.compiler.getClass());
```

Comment:
Since this codeblock depends on compiler you can move the code to Compiler (for example Compiler#processAnnotations()) and other compilers can overload it with a NOOP.

Reply:
The Compiler API is defined in a separate ECJ plugin org.eclipse.jdt.core.compiler.batch. I want to maintain a narrow focus for this pull request, and prefer not to modify too much bundles. Also ECJ has no concept of annotation processing, it's not good to move it to ECJ compiler API.

Reply:
how about ICompilerFactory.processAnnotations(...)?

Reply:
It would move complex logic, such as CompilationParticipant, into the ICompilerFactory, which seems overly designed. Instead, I'm thinking if we can introduce a flag such as ICompilerFactory.canCompilerSupportAnnotationProcessing to control whether the AbstractImageBuilder should skip annotation processing.

Reply:
I think the current code is clearer than the proposals: in practice annotation processors as CompilationParticipants are only necessary for ECJ, and don't make sense for other known implementation. The current code makes it really clear and easy to understand. Adding some indirection will make things more complicated to grok.
So IMO, we should stick to `isECJUsed`, or maybe just

```java
if (this.compiler instanceof Compiler) { // annotation processors as compilation participants are only relevant to ECJ Compiler
	// even if no files have annotations, must still tell every annotation processor in case the file used to have them
	for (CompilationParticipant participant : this.javaBuilder.participants)
		if (participant.isAnnotationProcessor())
			participant.processAnnotations(results);
}
```

Let's keep in mind that the code here is internal and we'd like to keep things simple for the moment. Maybe a 3rd implementation will allow to sort things out better now, but with only 2 implementations, we won't be able to ensure we make the perfect final choice anyway.